selfdrive/locationd/locationd.cc

```diff
@@ -243,9 +243,24 @@ void Localizer::handle_sensors(double current_time, const capnp::List<cereal::Se
   }
 }
 
+void Localizer::input_fake_gps_observations(double current_time) {
+  // This is done to make sure that the error estimate of the position does not blow up
+  // when the filter is in no-gps mode or in a tunnel
+  VectorXd current_x = this->kf->get_x();
+  
+  VectorXd ecef_pos = current_x.segment<3>(0);
+  VectorXd ecef_vel = current_x.segment<3>(7);
+  MatrixXdr ecef_pos_R = this->kf->get_fake_gps_pos_cov();
+  MatrixXdr ecef_vel_R = this->kf->get_fake_gps_vel_cov();
+  
+  this->kf->predict_and_observe(current_time, OBSERVATION_ECEF_POS, { ecef_pos }, { ecef_pos_R });
```

Comment:
This isn't ideal, you want to predict, then get the values then observe those. This pressures the position and velocity to stay the same, not move as predicted by the other sensors.

Reply:
If the observation noise is high enough, this is not true right? The more certain observations (from other sensors) have a higher say in the estimate than the low certainty sensor (as seen in the experiments).

Reply:
The effect may be minimal, but it's still more correct to get the predicted state.

You can just expose this function to do that.
https://github.com/commaai/rednose/blob/5b526a8e00bdc1c3922be470af1602cf9dc72dde/rednose/helpers/ekf_sym.cc#L197

Reply:
Yo, added this.